django/db/models/functions/text.py

```diff
@@ -204,6 +207,117 @@ def as_sqlite(self, compiler, connection, **extra_context):
+        if vendor in self.default_flags:
+            flags.value = self.default_flags[vendor] + flags.value
+
+        # Wrap pattern in group if required and increment backreferences.
+        if pattern.value and vendor in self.wrap_pattern and not extra_context.get('no_wrap'):
+            pattern.value = '(%s)' % re.sub(
+                r'\\([0-9]+)',
+                lambda m: r'\%d' % (int(m.group(1)) + 1),
+                pattern.value,
+            )
+
+        # Remove duplicate flags preserving the last occurrence of each to
+        # ensures that the last flag is preferred if any are contradictory.
+        # PY38: list() is no longer required to use reversed() on dict().
+        if flags and flags.value:
+            flags.value = ''.join(reversed(list(dict.fromkeys(reversed(flags.value)))))
+
+        # Resolve case-sensitivity flags preferring the last flag specified.
+        if {'c', 'i'}.issubset(flags.value):
+            remove = 'i' if flags.value.index('c') > flags.value.index('i') else 'c'
+            flags.value = flags.value.replace(remove, '')
+
+        if vendor in self.flag_mapping:
+            flags.value = flags.value.translate(self.flag_mapping[vendor])
+
+        if vendor in {'mysql', 'oracle'}:
+            position = Value(1)
+            occurrence = Value(0 if 'g' in flags.value else 1)
+            flags.value = flags.value.replace('g', '')
+            extra += (position, occurrence, flags)
+        elif vendor in self.inline_flags:
+            if flags.value and pattern.value:
+                # FIXME: Can this hack be improved?
+                if '-i' in flags.value:
+                    pattern.value = '(?-i)' + pattern.value
+                    flags.value = flags.value.replace('-i', '')
+                pattern.value = ('(?%s)' % flags.value.replace('g', '')) + pattern.value
+        else:
+            extra += (flags,)
+
+        self.set_source_expressions([expression, pattern, *extra])
+        return super().as_sql(compiler, connection, **extra_context)
+
+
+class RegexpReplace(RegexpFlagMixin, Func):
+    function = 'REGEXP_REPLACE'
+    inline_flags = {'mariadb'}
+
+    def __init__(self, expression, pattern, replacement=Value(''), flags=Value(''), **extra):
+        super().__init__(expression, pattern, replacement, flags, **extra)
+
+
+class RegexpStrIndex(RegexpFlagMixin, Func):
+    function = 'REGEXP_INSTR'
+    inline_flags = {'mariadb'}
+
+    def __init__(self, expression, pattern, flags=Value(''), **extra):
+        super().__init__(expression, pattern, flags, **extra)
+
+    def as_postgresql(self, compiler, connection, **extra_context):
+        # FIXME: This emulated version doesn't handle NULL pattern correctly.
+        expression, pattern, flags = self.source_expressions.copy()
+        expr = RegexpSubstr(expression, pattern, flags, no_wrap=True)
+        expr = StrIndex(expression, Coalesce(expr, Value('<<fail>>')))
```

Comment:
Emulating `REGEXP_INSTR` in PostgreSQL is tough. Unfortunately empty string has an index of 1 so we need to coalesce to some other sentinel value that will not match to get the expected index of 0 for no match. This breaks for `NULL` being passed to `pattern`. We might be able to solve this by wrapping with `Case`.

Reply:
It should be possible to do `CASE WHEN <pattern> IS NULL THEN NULL ELSE <expr> END` in PostgreSQL, but I can't work out if this is possible with Django's `Case` and `When`...

Reply:
Am now thinking that I'll probably just skip the test for `NULL` being passed to `pattern` and chalk it up as a wart. I'm having to do that for Oracle anyway for `REGEXP_REPLACE` as it treats `NULL` as `''` for `pattern` there, returning the original string instead of `NULL`.